File: 2015/day-02/day-02-i-was-told-there-would-be-no-math.rs
use std::fs;

#[derive(Debug)]
struct Package {
    width: u32,
    height: u32,
    length: u32,
}

impl Package {
    fn new(sides: Vec<u32>) -> Self {
        Self {
            width: sides[0],
            height: sides[1],
            length: sides[2],
        }
    }

    fn surface_area(&self) -> u32 {
        2 * self.length * self.width + 2 * self.width * self.height + 2 * self.height * self.length
    }

    fn volume(&self) -> u32 {
        self.length * self.width * self.height
    }

    fn smallest_side_area(&self) -> u32 {
        self.width * self.height
    }

    fn smallest_face_circum(&self) -> u32 {
        2 * self.width + 2 * self.height
    }
}

fn main() {
    let filepath = "./day-02-data.txt";
    let input = parse_input(filepath);
    let mut total_wrapping_paper = 0;
    let mut total_ribbon = 0;
    for package in input {
        total_wrapping_paper += package.surface_area() + package.smallest_side_area();

        total_ribbon += package.smallest_face_circum() + package.volume();
    }

    println!("Part one: {}", total_wrapping_paper);
    println!("Part two: {}", total_ribbon);
}

fn parse_input(filepath: &str) -> Vec<Package> {
    let mut packages = Vec::new();

    let file = fs::read_to_string(filepath).expect("Cannot read file");

    for line in file.split('\n') {
        // filter out empty line at the end
        if !line.is_empty() {
            let mut split: Vec<u32> = line
                .split('x')
                .map(|x| match x.parse::<u32>() {
                    Ok(y) => y,
                    Err(_e) => 0,
                })
                .collect();
            split.sort();
            packages.push(Package::new(split));
        }
    }

    packages
}

File: 2015/day-02/day-02-i-was-told-there-would-be-no-math.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	// read each line in file
	f, _ := os.Open("./day-02-data.txt")
	fs := bufio.NewScanner(f)
	fs.Split(bufio.ScanLines)
	var total_paper_area int
	var total_ribbon_length int
	for fs.Scan() {
		var gift_dimensions []int
		for _, d := range strings.Split(fs.Text(), "x") {
			dimension, _ := strconv.Atoi(d)
			gift_dimensions = append(gift_dimensions, dimension)
		}

		sort.Slice(gift_dimensions, func(i, j int) bool {
			return gift_dimensions[i] < gift_dimensions[j]
		})

		total_paper_area += side_surface_area(gift_dimensions[0], gift_dimensions[1]) + total_surface_area(gift_dimensions[0], gift_dimensions[1], gift_dimensions[2])
		total_ribbon_length += side_perimeter(gift_dimensions[0], gift_dimensions[1]) + volume(gift_dimensions[0], gift_dimensions[1], gift_dimensions[2])
	}

	fmt.Println("Part one:", total_paper_area)
	fmt.Println("Part two:", total_ribbon_length)
}

func side_surface_area(w, h int) int {
	return w * h
}

func total_surface_area(w, h, l int) int {
	return 2*side_surface_area(w, h) + 2*side_surface_area(h, l) + 2*side_surface_area(w, l)
}

func side_perimeter(w, h int) int {
	return 2*w + 2*h
}

func volume(w, h, l int) int {
	return w * h * l
}
